src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -1284,5 +1284,51 @@ public void ChainWithMaxReorgPlusExtraHeadersIsCalled_AnotherChainIsPresented_Co
                 }
             }
         }
+
+        /// <summary>
+        /// Issue 33 @  We receive headers message
+        /// (first header in the message is HEADERS_START and last is HEADERS_END).
+        /// AV = assume valid header, CP1,CP2 - checkpointed headers. '---' some headers.
+        /// HEADERS_START--CP1----AV----HEADERS_END---CP2---
+        /// Check that headers till AV(including AV) are marked for download.
+        /// Headers after AV are not marked for download.
+        /// </summary>
+        [Fact]
+        public void ConnectHeaders_AssumeValidBetweenTwoCheckPoints_DownloadUpToIncludingAssumeValid()
+        {
+            const int initialChainSize = 5;
+            const int chainExtension = 40;
+            const int assumeValidHeaderHeight = 30;
+
+            TestContext testContext = new TestContextBuilder().WithInitialChain(initialChainSize).UseCheckpoints(true).Build();
+            ChainedHeaderTree chainedHeaderTree = testContext.ChainedHeaderTree;
+            ChainedHeader initialChainTip = testContext.InitialChainTip;
+            ChainedHeader chainTip = testContext.ExtendAChain(chainExtension, initialChainTip);
+
+            // Assume valid header at h30.
+            ChainedHeader assumeValidHeader = chainTip.GetAncestor(assumeValidHeaderHeight);
+            testContext.ConsensusSettings.BlockAssumedValid = assumeValidHeader.HashBlock;
+
+            List<BlockHeader> listOfChainHeaders = testContext.ChainedHeaderToList(chainTip, initialChainSize + chainExtension);
+
+            // Two checkpoints at h18 and h45.
+            const int firstCheckpointHeight = 18;
+            const int secondCheckpointHeight = 45;
+            var checkpoint1 = new CheckpointFixture(firstCheckpointHeight, listOfChainHeaders[firstCheckpointHeight - 1]);
+            var checkpoint2 = new CheckpointFixture(secondCheckpointHeight, listOfChainHeaders[secondCheckpointHeight - 1]);
+            testContext.SetupCheckpoints(checkpoint1, checkpoint2);
+
+            // Present chain h1->h40 covering first checkpoint at h18 and assume valid at h30.
+            // Exclude second checkpoint at h45.
+            listOfChainHeaders = listOfChainHeaders.Take(40).ToList();
+            ConnectNewHeadersResult connectedHeadersResultNew = chainedHeaderTree.ConnectNewHeaders(1, listOfChainHeaders);
+
+            // From initialised chain up to and including assume valid (h6->h30) are marked for download.
+            // Headers after assume valid (h31->h40) are not marked for download.
+            IEnumerable<BlockHeader> extendedChain = listOfChainHeaders.Skip(initialChainSize);
+            
+            Assert.Equal(extendedChain.First(), connectedHeadersResultNew.DownloadFrom.Header);
+            Assert.Equal(assumeValidHeader.Header, connectedHeadersResultNew.DownloadTo.Header);
```

Comment:
All good here, just please include checking of block data availability state for all these headers that are marked for download as well as for those that are not marked for download after assumed valid block.

Reply:
Love your use of CTRL+K+D

Reply:
Fixed!